Update API to take ownership where needed

ionut-arm opened the pull request:
Updating functions and methods in the API (particularly for `Context`
methods) where we take a reference and then clone the object, instead of
taking ownership directly, letting the user clone whenever needed.

This includes updates to the tests and documentation.

Signed-off-by: Ionut Mihalcea <ionut.mihalcea@arm.com>

wiktor-k commented on tss-esapi/src/context/tpm_commands/non_volatile_storage.rs:
```diff
@@ -39,8 +39,8 @@ impl Context {
                 self.optional_session_1(),
                 self.optional_session_2(),
                 self.optional_session_3(),
-                &auth.cloned().unwrap_or_default().into(),
```
Interesting that this function was called `cloned` with **d** :thinking:

Superhepper replied:
I think it is to distinguish from cloning the reference or cloning the referenced data.

wiktor-k replied:
Understood, thanks for context. Since now this has been removed maybe that function is no longer needed? Just asking as I never used it :)

wiktor-k replied:
`cloned`... but after a careful study of other PRs I see it in a few places here and there... I wonder why haven't I seen it earlier :thinking: It's probably all OK now :+1:

wiktor-k reviewed (approved):
Looks great :+1: Thanks for going through the entire code and adjusting :)

ionut-arm commented:
I've rebased and also found a few more places that were doing extra clones, so changed that as well. No need to constantly re-review unless you're particularly keen. It'll be necessary once the other PRs are out of the way and I've rebased on top of them (I'll mark this as ready for review)

ionut-arm commented on tss-esapi/src/structures/pcr/select.rs:
```diff
@@ -27,10 +27,10 @@ pub struct PcrSelect {
     /// Creates a new PcrSelect
-    pub fn new(size_of_select: PcrSelectSize, pcr_slots: &[PcrSlot]) -> Self {
+    pub fn new(size_of_select: PcrSelectSize, mut pcr_slots: Vec<PcrSlot>) -> Self {
         PcrSelect {
             size_of_select,
-            selected_pcrs: pcr_slots.iter().cloned().collect(),
+            selected_pcrs: pcr_slots.drain(..).collect(),
```
I wasn't sure about this one, whether `Vec` is a fair replacemenet or whether I should use something else, maybe something generic?

Superhepper replied:
hrrmm, that is weird it is only a value.. should be no need to clone it.

ionut-arm replied:
Before we used to get `&[PcrSlot]` - so we had to clone the stuff behind the reference.

Superhepper replied:
This might work:
```rust
    /// Creates a new PcrSelect
    pub fn new(size_of_select: PcrSelectSize, pcr_slots: &[PcrSlot]) -> Self {
        PcrSelect {
            size_of_select,
            selected_pcrs: pcr_slots
                .iter()
                .fold(BitFlags::<PcrSlot>::empty(), |mut acc, &v| {
                    acc.insert(v);
                    acc
                }),
        }
    }
```

ionut-arm replied:
True, but that's essentially doing a (potentially more expensive) version of `cloned`, by hand

ionut-arm replied:
Oh, that's literally what's happening under the hood with the `.collect()` implementations: https://docs.rs/enumflags2/0.7.3/src/enumflags2/lib.rs.html#853

Superhepper replied:
Yeah, it is a set under the hood. Would be nice to be able to it without cloning. Either using from_iterator or collect.

Superhepper replied:
```
    /// Creates a new PcrSelect
    pub fn new(size_of_select: PcrSelectSize, pcr_slots: &[PcrSlot]) -> Self {
        PcrSelect {
            size_of_select,
            selected_pcrs: BitFlags::<PcrSlot>::from_iter(pcr_slots.iter().copied()),
        }
    }
```

```
    /// Creates a new PcrSelect
    pub fn new(size_of_select: PcrSelectSize, pcr_slots: &[PcrSlot]) -> Self {
        PcrSelect {
            size_of_select,
            selected_pcrs: pcr_slots.iter().copied().collect(),
        }
    }
```

Both works and are better then the previous.

Superhepper reviewed (approved):
Nice thanks for this.

ionut-arm commented:
@Superhepper - should I wait for your other PRs to merge this?
